Add FrsClient documentation to FF.com (#6736)
microsoft/FluidFramework: docs/content/docs/deployment/azure-frs.md

---

> title: Connect to a Fluid Service in Azure with FrsClient

Reviewer: ```suggestion
title: Connect to a Azure Fluid Relay service
```

menuPosition: 2
---

# Introduction

> Fluid provides an implementation of its backend service, known as the Azure Fluid Relay Service (FRS), for clients to connect to. As part of the onboarding process, you will be provided a tenant ID, a tenant key, and orderer and storage URLs that will be used to connect your application to this service instance. The framework also provides a client-side package, `frs-client`, that will take in these configuration values to quickly allow you set up your collaborative application against the service instance. This package handles the logic of connecting your [Fluid Container](../build/containers.md) to the service while keeping the container object itself service-agnostic. You can use one instance of this client to manage multiple containers. 

Reviewer: Nit: The official marketing direction is to use a small s for service. `Azure Fluid Relay service`

> Fluid provides an implementation of its backend service, known as the Azure Fluid Relay Service (FRS), for clients to connect to. As part of the onboarding process, you will be provided a tenant ID, a tenant key, and orderer and storage URLs that will be used to connect your application to this service instance. The framework also provides a client-side package, `frs-client`, that will take in these configuration values to quickly allow you set up your collaborative application against the service instance. This package handles the logic of connecting your [Fluid Container](../build/containers.md) to the service while keeping the container object itself service-agnostic. You can use one instance of this client to manage multiple containers. 

Reviewer: FRS is an internal term and while we can keep it to refer to the class names and packages for now we should use Azure Fluid Relay proper when talking about the service.

> Fluid provides an implementation of its backend service, known as the Azure Fluid Relay Service (FRS), for clients to connect to. As part of the onboarding process, you will be provided a tenant ID, a tenant key, and orderer and storage URLs that will be used to connect your application to this service instance. The framework also provides a client-side package, `frs-client`, that will take in these configuration values to quickly allow you set up your collaborative application against the service instance. This package handles the logic of connecting your [Fluid Container](../build/containers.md) to the service while keeping the container object itself service-agnostic. You can use one instance of this client to manage multiple containers. 

Author: Got it! Thanks for the clarification! Will refer to it as Azure Fluid Relay service!

Let's take a look at how to go about using the `frs-client` in your app!

# Connecting to the Service

> To connect to our FRS instance, we first need to instaniate our `FrsClient`. This takes in, as configuration parameters, the tenant ID, orderer, and storage URLs that were provided as part of the FRS onboarding process. It also requires a token provider to generate the JWT token that will be used to authorize the current user against the service. The `InsecureTokenProvider` should only be used for testing purposes as it exposes the tenant key secret in your client-side code bundle. This should be replaced with an implementation of `ITokenProvider` that fetches the token from your own backend service that is responsible for signing it with the tenant key. 

Reviewer: I like the warning callout on the token provider (y)

> To connect to our FRS instance, we first need to instaniate our `FrsClient`. This takes in, as configuration parameters, the tenant ID, orderer, and storage URLs that were provided as part of the FRS onboarding process. It also requires a token provider to generate the JWT token that will be used to authorize the current user against the service. The `InsecureTokenProvider` should only be used for testing purposes as it exposes the tenant key secret in your client-side code bundle. This should be replaced with an implementation of `ITokenProvider` that fetches the token from your own backend service that is responsible for signing it with the tenant key. 

Reviewer: Add a few nits to Tyler's suggested edit.

> To connect to our FRS instance, we first need to instaniate our `FrsClient`. This takes in, as configuration parameters, the tenant ID, orderer, and storage URLs that were provided as part of the FRS onboarding process. It also requires a token provider to generate the JWT token that will be used to authorize the current user against the service. The `InsecureTokenProvider` should only be used for testing purposes as it exposes the tenant key secret in your client-side code bundle. This should be replaced with an implementation of `ITokenProvider` that fetches the token from your own backend service that is responsible for signing it with the tenant key. 

Author: Thank you for the suggestions and the link for guidance! I've incorporate the commit in and had move the snippet into its own file into _includes/token-provider-warning.md
When I refer to it in this file using `{{% include file="_includes/token-provider-warning.md" %}}`, I see its bringing in the content but then it loses the danger formatting.
![image](https://user-images.githubusercontent.com/7992711/126245672-08aa10a5-4245-4100-b01e-79a2f6e663df.png)

The formatting shows up when I have the content inline. Any hints for how to have the danger tag picked up when using the reusable snippet?

> ```typescript
> const config: FrsConnectionConfig = {

Reviewer: Let's stick with javascript only for code snippets. Typescript types clutter the goodness.

```suggestion
```javascript
const config = {
```

    tenantId: "myFrsTenantId",

>     // IMPORTANT: this token provider is suitable for testing ONLY. It is NOT secure.

Reviewer: Let's delete this comment since it clutters the object and we are adding some clear callout warnings below around InsecureTokenProvider.

```suggestion
```

>     tenantId: "myFrsTenantId",
>     // IMPORTANT: this token provider is suitable for testing ONLY. It is NOT secure.
>     tokenProvider: new InsecureTokenProvider("myFrsTenantKey", { id: "UserId", name: "Test User" }),
>     orderer: "https://myFrsOrdererUrl",
>     storage: "https://myFrsStorageUrl",

Reviewer: Scrubbing frs

```suggestion
    tenantId: "myTenantId",
    // IMPORTANT: this token provider is suitable for testing ONLY. It is NOT secure.
    tokenProvider: new InsecureTokenProvider("myTenantKey", { id: "UserId", name: "Test User" }),
    orderer: "https://myOrdererUrl",
    storage: "https://myStorageUrl",
```

}

const client = new FrsClient(config);
```

Now that we have an instance of our client, we can start using it to create new or load existing Fluid containers!

> 

Reviewer: Adding a stub for TokenProviders. They definitely deserve a sub-section of their own. The first thing I thought of after reading the warning was, "well if I can't use this, what do I use?

```suggestion

##  Token Providers

```

Author: Yep I was thinking that section would go in with the instructions for securely setting up the token provider with the Azure Function but its a good idea to stub it out!

# Managing Containers

The `FrsClient` provides two functions on its API to create and get containers respectively. They both take in two parameters:
- A container config that defines the ID of the container and an optional entrypoint for logging
- A container schema that defines the DDSes and data objects that this container will hold

> ```typescript

Reviewer: ```suggestion
```javascript
```

const schema = {
    name: "my-container",
    initialObjects: {
        /* ... */
    },
    dynamicObjectTypes: [ /*...*/ ],

> }

Reviewer: TS does not infer type here from below but will create a type based on the provided properties and types. It will give an error below if the schema doesn't fit the type.

Let's keep types out of the snippets to keep them clean :)

> }

Author: Converted all the snippets to JS

const frsClient = new FrsClient(config);
await frsClient.createContainer({ id: "_unique-id_" }, schema);
const { fluidContainer, containerServices } = await frsClient.getContainer({ id: "_unique-id_" }, schema);
```

> The ID that is being passed in to the container config can be thought of as the "filename" for the item holding the data for this container on the FRS backend. Any client that wants to join the same collaborative session just needs to call `getContainer` with the same container ID.

Reviewer: Let's not introduce a file type metaphor since it's really only useful if the developer is coming in with files as their default paradigm.

```suggestion
The `id` being passed into the container config is a unique identifier to a container instance. Any client that wants to join the same collaborative session just needs to call `getContainer` with the same container `id`.
```

> For the further information on how to start recording logs being emitted by Fluid, please see [Telemetry](../testing/telemetry.md)

Author: Updated all links to use hugo! Thank you for providing the example, it was super helpful!

The container being fetched back will hold the `initialObjects` as defined in the container schema. Please read [Data modeling](../build/data-modeling.md) to see how to establish the schema and use the `fluidContainer` object.

> # Getting Audience Details

Reviewer 2: ```suggestion
# Using audience to manage active users
```

How do you think this section relates/should relate to the "User presence and audience" article?

Author: I was interpreting this sections as a way of understanding how to use the `FrsAudience` object specifically so that we can get the specific user details for FRS back and provide an example of how to render them out. Whereas "User presence and audience" would provide the general guidance for how to use the audience details from any client to manage the roster of members within the app state

Reviewer 2: Makes sense. When we have both published we can tweak each one to make things clearer if needed.

> Both calls to `createContainer` and `getContainer` return an `FrsResources` object that holds the Fluid container that we were discussing above as well as a `containerServices` object. Whereas the container itself will always stay the same regardless of which service it is being connected, the `containerServices` hold values that are specific to the FRS service. Within this object, we will find an `audience` value that can be used to manage the roster of users that are currently collaborating in the container.

Reviewer 2: ```suggestion
Both calls to `createContainer` and `getContainer` return an `FrsResources` object that holds the Fluid container described above as well as a `containerServices` object.

The Fluid container is service-agnostic. If you were to retrieve a Fluid container from a different Fluid service -- not FRS -- the code you wrote against that container using FRS is portable between services.

The `containerServices` object contains data that is specific to the FRS service. This object contains an `audience` value that can be used to manage the roster of users that are currently collaborating in the container.
```

The container service-agnostic thing is _really_ tricky to communicate! I don't love my changes. Maybe just using the class name instead would help?

Reviewer: ```suggestion
Calls to `createContainer` and `getContainer` return an `FrsResources` object that contains a `FluidContainer` -- described above -- and a `containerServices` object.

The `containerServices` object contains data that is specific to the Azure Fluid Relay service. This object contains an `audience` value that can be used to manage the roster of users that are currently connected to the container.
```

Author: Thank you for the suggestions here @tylerbutler and @skylerjokiel! It's definitely sounding a lot clearer with these recommendations and I think conveying the "service-agnostic" aspect is going to be key for ppl to understand so they know they can keep moving their container between services. Just made a small addition to specify that this means that they don't have to change any of their interactions with DDSes and data objects that are within their container

```suggestion
Calls to `createContainer` and `getContainer` return an `FrsResources` object that contains a `FluidContainer` -- described above -- and a `containerServices` object.

The `FluidContainer` contains the Fluid data model and is service-agnostic. Any code you write against this container object returned by the `FrsClient` is reusable with the client for another service. An example of this is if you prototyped your scenario using `TinyliciousClient`, then all of your code interacting with the Fluid DDSes and data objects within the container can be reused when moving to using `FrsClient`.

The `containerServices` object contains data that is specific to the Azure Fluid Relay service. This object contains an `audience` value that can be used to manage the roster of users that are currently connected to the container.
```

> Both calls to `createContainer` and `getContainer` return an `FrsResources` object that holds the Fluid container that we were discussing above as well as a `containerServices` object. Whereas the container itself will always stay the same regardless of which service it is being connected, the `containerServices` hold values that are specific to the FRS service. Within this object, we will find an `audience` value that can be used to manage the roster of users that are currently collaborating in the container.

Reviewer 2: Is this deliberately not a tuple?

Reviewer: It was originally a tuple but was changed to an object. I'm still not convinced that an object is the right pattern here and I really dislike that we have to define the return object type. @ChumpChief / @sumedhb1995 who drove this change.

Author: There were multiple requests from consumers of the create/getContainer call who found that indexing the returned value like `const fluidContainer = (await client.createContainer(....))[0]` when they only need one of the two values was affecting code readability and keeping it as a tuple wasn't actually restricting people's ability to add on additional returned values in the array.

Reviewer 2: But wouldn't the idiomatic code look like `const [fluidContainer] = (await client.createContainer(....))`? To be clear the object is OK with me I just want to better understand the reasoning.

Author: Correct, but I believe the theory is that the developer could also technically write `const [containerServices, fluidContainer] = (await client.createContainer(....))` and they wouldn't see any error, even though they've swapped their returned values around. Or they could do `const fluidContainer = (await client.createContainer(....))[1]` and again get the wrong value. It relies on the developer and how they choose to parse the two values out. Whereas with the object, they need to explicitly get the `fluidContainer` or `containerServices` values from it

> `audience` provides two callbacks that will return `FrsMember` objects that have a user ID and user name:

Reviewer 2: I was immediately confused that it provides callbacks... Maybe start with a brief explanation that the audience changes as users join and leave, hence your code is handling those events.

Reviewer: They are not callbacks though. They are just functions.

Reviewer 2: Good point. I think starting with "here's what audience is and how it's managed" will help. Maybe moving the code sample above the explanation would help too.

Author: Great suggestions! Moved the code sample up and changed "callbacks" to be "functions". Also, I added in a few lines stating that each call to `getMember` would reflect the new values after the events fire. And moved up the events description to be right after the functions description so its easy to link the two together.

> - `getMembers` returns a map of all the users that are currently in the Fluid session

Reviewer 2: Showing some sample data would really help communicate what this looks like. Well, it would help me t least. 😂

Author: Totally fair, added in the sample data for what is returned in an `FrsMember`!

> - `getMembers` returns a map of all the users that are currently in the Fluid session
> - `getMyself` returns the current user on this client

Reviewer 2: ```suggestion
- `getMembers` returns a map of all the users that are currently in the Fluid session.
- `getMyself` returns the current user on this client.
```

nit: Fluid session is a new term, I think. aybe connected to the container? Connected to the service?

Reviewer: Let's use `container` as the only word we use. no session, no document.

```suggestion
- `getMembers` returns a map of all the users connected to the container.
- `getMyself` returns the current user on this client.
```

> Alongside the user ID and name, `FrsMember` objects also hold an array of `connections`. If the user is logged into the session with only one client, `connections` will only have one value in it with the ID of the client and if is in read/write mode. However, if the same user is logged in from multiple clients, `connections` here will hold multiple values for each client.

Reviewer 2: Sample data would help here too.

Reviewer 2: Actually, if you have sample data earlier, you can refer back to it here. Then the words here can focus on explaining connections specifically.

Author: Great point!! I added in a sample FrsMember object with two different clients connected and then reworded this to instead walk through the object and use it to explain what the `connections` value means

`audience` also emits events for when the roster of members changes. `membersChanged` will fire for any roster changes, whereas `memberAdded` and `memberRemoved` will fire for their respective changes with the `clientId` and `member` values that have been modified.

> These callbacks and events can be combined to keep an updated view that represents the members in the current session.

Reviewer 2: ```suggestion
These callbacks and events can be combined to present a real-time view of the users in the current session.
```

``` typescript
const { audience } = containerServices;
const audienceDiv = document.createElement("div");

const onAudienceChanged = () => {
    const members = audience.getMembers();
    const self = audience.getMyself();

>     const memberNames: string[] = [];

Reviewer 2: nit: I like types explicitly here too.

Reviewer: ```suggestion
const onAudienceChanged = () => {
    const members = audience.getMembers();
    const self = audience.getMyself();
    const memberNames = [];
```

    members.forEach((member) => {
        if (member.userId !== self?.userId) {
            memberNames.push(member.userName);
        }
    });
    audienceDiv.innerHTML = `
        Current User: ${self?.userName} <br />
        Other Users: ${memberNames.join(", ")}
    `;
};

onAudienceChanged();
audience.on("membersChanged", onAudienceChanged);
```

> Every time the `membersChanged` events gets fired, we fetch the new member roster and update the view accordingly.

Reviewer 2: ```suggestion
Every time the `membersChanged` event is sent, the new member roster is fetched and the view is updated accordingly.
```
